File: leetcode/Python/0016-m-3sum-closest.py
"""
Sort and use two pointers
Time: O(n^2)
Space: O(1)
"""
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        minDiff = float('inf')
        nums = sorted(nums)
        for i in range(len(nums) - 2):
            left, right = i + 1, len(nums) - 1
            while left < right:
                curSum = nums[i] + nums[left] + nums[right]
                curDiff = abs(curSum - target)
                if curDiff < minDiff:
                    minDiff = curDiff
                    res = curSum
                if curSum > target:
                    right -= 1
                elif curSum < target:
                    left += 1
                else:
                    return target
        return res

File: leetcode/C++/0016-m-3sum-closest.cpp
// Time: O(N^2)
// Space: O(logN)
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int best = 1e7;

        auto update = [&](int cur) {
            if (abs(cur - target) < abs(best - target)) {
                best = cur;
            }
        };

        for (int i = 0; i < n; ++i) {
            if (i > 0 && nums[i] == nums[i-1])
                continue;
            int j = i + 1, k = n - 1;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == target)
                    return target;
                update(sum);
                if (sum > target) {
                    --k;
                    while (j < k && nums[k] == nums[k + 1])
                        --k;
                } else {
                    ++j;
                    while (j < k && nums[j] == nums[j-1])
                        ++j;
                }
            }
        }
        return best;
    }
};
